### PR TITLE
[SPARK-18884][SQL] Throw an exception in compile time if Array[_] used in ScalaUDF

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr modified code to throw an exception in compile time if Array[_] used in the arguments of `ScalaUDF`. Currently, a query below throws an exception in runtime if we use the type in `ScalaUDF`;

```
scala> import org.apache.spark.sql.execution.debug._
scala> Seq((0, 1)).toDF("a", "b").select(array($"a", $"b").as("ar")).write.mode("overwrite").parquet("/Users/maropu/Desktop/data/")
scala> val df = spark.read.load("/Users/maropu/Desktop/data/")
scala> val df = Seq((0, 1)).toDF("a", "b").select(array($"a", $"b").as("ar"))
scala> val testArrayUdf = udf { (ar: Array[Int]) => ar.sum }
scala> df.select(testArrayUdf($"ar")).show

Caused by: java.lang.ClassCastException: scala.collection.mutable.WrappedArray$ofRef cannot be cast to [I
  at $anonfun$1.apply(<console>:23)
  at org.apache.spark.sql.catalyst.expressions.ScalaUDF$$anonfun$2.apply(ScalaUDF.scala:89)
  at org.apache.spark.sql.catalyst.expressions.ScalaUDF$$anonfun$2.apply(ScalaUDF.scala:88)
  at org.apache.spark.sql.catalyst.expressions.ScalaUDF.eval(ScalaUDF.scala:1069)
  ... 99 more
```

## How was this patch tested?
Added tests in `DataFrameSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
I'm looking for another approach not to break backward compatibility...

**Comment by maropu** [comment]
@dongjoon-hyun Could you take time to review this before committers do? Thanks!

**Comment by dongjoon-hyun** [comment]
Sure, @maropu . I'll do that tomorrow morning (PST).

**Comment by maropu** [comment]
many thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 1]
```diff
@@ -84,7 +86,9 @@ case class ScalaUDF(
     case 1 =>
       val func = function.asInstanceOf[(Any) => Any]
       val child0 = children(0)
-      lazy val converter0 = CatalystTypeConverters.createToScalaConverter(child0.dataType)
+      lazy val converter0 = inputConverters.map {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: Also, please update the template [comment](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala#L52-L71) and follow the similar syntax.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @maropo . This part consists of many generated codes. So, could you update the template in `comment` together and update the newly added code to use the similar generated syntax?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 2]
```diff
@@ -1428,6 +1428,134 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
...
+            udf { (ar1: Seq[Int], ar2: Seq[Int], ar3: Seq[Int], ar4: Seq[Int], ar5: Seq[Int], ar6: Seq[Int]) => (ar1 ++ ar2 ++ ar3 ++ ar4 ++ ar5 ++ ar6).sum },
+            udf { (ar1: Seq[Int], ar2: Seq[Int], ar3: Seq[Int], ar4: Seq[Int], ar5: Seq[Int], ar6: Seq[Int], ar7: Seq[Int]) => (ar1 ++ ar2 ++ ar3 ++ ar4 ++ ar5 ++ ar6 ++ ar7).sum },
+            udf { (ar1: Seq[Int], ar2: Seq[Int], ar3: Seq[Int], ar4: Seq[Int], ar5: Seq[Int], ar6: Seq[Int], ar7: Seq[Int], ar8: Seq[Int]) => (ar1 ++ ar2 ++ ar3 ++ ar4 ++ ar5 ++ ar6 ++ ar7 ++ ar8).sum },
+            udf { (ar1: Seq[Int], ar2: Seq[Int], ar3: Seq[Int], ar4: Seq[Int], ar5: Seq[Int], ar6: Seq[Int], ar7: Seq[Int], ar8: Seq[Int], ar9: Seq[Int]) => (ar1 ++ ar2 ++ ar3 ++ ar4 ++ ar5 ++ ar6 ++ ar7 ++ ar8 ++ ar9).sum },
+            udf { (ar1: Seq[Int], ar2: Seq[Int], ar3: Seq[Int], ar4: Seq[Int], ar5: Seq[Int], ar6: Seq[Int], ar7: Seq[Int], ar8: Seq[Int], ar9: Seq[Int], ar10: Seq[Int]) => (ar1 ++ ar2 ++ ar3 ++ ar4 ++ ar5 ++ ar6 ++ ar7 ++ ar8 ++ ar9 ++ ar10).sum }
+          )
+        ).map { case (udf1, udf2, udf3, udf4, udf5, udf6, udf7, udf8, udf9, udf10) =>
+          val arVal = functions.array(lit(1), lit(1))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: you mean this kind of this?: 
```

val testUdf = udf { (ar: Array[Long]) => ar.sum }
val df = spark.range(10).select(array('id, 'id).as("arVal"))
checkAnswer(df.select(udf1(arVal)), Row(2) :: Nil)
```

Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: +1. Yes.

Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: okay!

**Comment by maropu** [comment]
@dongjoon-hyun okay, I applied your comments into this pr. Could you check again to satisfy your intentions?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala (hunk 1 above)
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: Hi, I think you missed this comment.

**Comment by dongjoon-hyun** [comment]
Hi, @maropu .

In the current master with your example, we can do the following. How do you think about this?

```scala
scala> import scala.collection.mutable.WrappedArray
import scala.collection.mutable.WrappedArray

scala> val testArrayUdf = udf { (ar: WrappedArray[Int]) => ar.sum }
testArrayUdf: org.apache.spark.sql.expressions.UserDefinedFunction = UserDefinedFunction(<function1>,IntegerType,Some(List(ArrayType(IntegerType,false))))

scala> df.select(testArrayUdf($"ar")).show
+-------+
|UDF(ar)|
+-------+
|      1|
+-------+
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala (hunk 1 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: oh, sorry. I'll do it soon.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: okay, fixed!

**Comment by maropu** [comment]
oh, yea. I didn't find that and I think it's a good point.
IMO `WrappedArray` is implicitly used inside for implicit conversions, so users do not use `WrappedArray` directly for UDFs in most cases.

Anyway, thanks alots for your reviews!

**Comment by dongjoon-hyun** [comment]
Sure, @maropu . `WrappedArray` is not documented for now.

Hi, @gatorsmile and @cloud-fan .
Could you review this PR?

**Comment by cloud-fan** [comment]
Well, it will be good if we can support `Array` in `ScalaUDF`, but it's not a big deal as users can easily do `udf { (seq: Seq[Int]) => val a = seq.toArray; // do anything you like with the array }`.

considering the size of this PR, I don't think it worth.

**Comment by maropu** [comment]
okay. But, if this issue finished, I'm planning to take SPARK-12823 in a similar way.
Do u think also it's not also worth trying struct? cc: @cloud-fan

**Comment by maropu** [comment]
The workaround @cloud-fan said is good to me though, IMO the most critical issue here is that this cast exception happens not in analysis phases but in runtime. So, at least I think we should modify code to throw an exception in analysis phases and the exception might include a message like "you should use `Seq[T]` instread of `Array[T]`". I think we could do this with less amount of code. Thought?

**Comment by cloud-fan** [comment]
SGTM

**Comment by maropu** [comment]
okay, I'll update this pr in that way, thanks!

**Comment by maropu** [comment]
@cloud-fan Could you give me more insights on this?

**Comment by maropu** [comment]
@cloud-fan ping

**Comment by maropu** [comment]
@cloud-fan ping

**Comment by maropu** [comment]
I'll close for now

**Pull request closed without merging**